cirq-core/cirq/protocols/decompose_protocol.py

```diff
@@ -138,6 +148,9 @@ def decompose(
         None, Exception, Callable[['cirq.Operation'], Optional[Exception]]
     ] = _value_error_describing_bad_operation,
     preserve_structure: bool = False,
+    qubit_manager: Optional['cirq.QubitManager'] = None,
+    constraint: DecomposeConstraints = lambda _: True,
+    objective: DecomposeObjective = DecomposeObjective.MINIMUM_T_COUNT,
```

Comment:
Another way of supporting the `objective` of decomposition would be to instantiate gates with different flags based on which decomposition should be preferred. For example: Assume you want a multi controlled X gate in your circuit and it supports 3 different decompositions based on whether you want to use clean ancillas, dirty ancillas or no ancillas. You could make this choice either  

a) when instantiating and inserting the gate in your circuit (eg: `MultiControlledX(args, version=NO_ANCILLA)`) or 
b) when calling `cirq.decompose(circuit, version= MINIMUM_CIRCUIT_WIDTH)`. 

(a) gives you more fine grained control over choosing a specific version of the decomposition for each composite gate in your circuit whereas (b) let's you make a global choice for the entire circuit. 

I think (a) is a lot more flexible since there might not be a single "version" option that makes sense for the entire circuit. And, if we do end up often in a situation where it makes sense to choose a global "version" for the entire circuit; we can always write a circuit transformer s.t. `minimum_t_count_circuit = cirq.map_operations(circuit, map_func=lambda op, _: op.with_version(MINIMUM_CIRCUIT_WIDTH)` and then `cirq.decompose(minimum_t_count_circuit)` would be equivalent to `cirq.decompose(circuit, version= MINIMUM_CIRCUIT_WIDTH)`. 



Hence, I think for now we should just add support for the `qubit_manager` keyword argument. What do you think?